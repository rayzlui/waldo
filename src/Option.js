import React from 'react'

class Options extends React.Component{
    constructor(props){
        super(props)
        this.generateSelectOptionsForGame = this.generateSelectOptionsForGame.bind(this)
    }

    generateSelectOptionsForGame(id){
        let options = []
        let tag
        let val
        let randomKey
        //random decides which select option is going to be the correct one.
        //if we're playing game we should ONLY USE KEYS THAT HAVE VALUES.
        let tagKeys = Object.keys(this.props.tags).map(x=>x/1)
        tagKeys.splice(tagKeys.indexOf(id),1)
        //remove id from the wrong options to prevent duplicate correct answer.
    
        let limit = tagKeys.length > 5? 4 : tagKeys.length//
        let random = Math.floor(Math.random()*(limit))
        for (var i = 0; i<=limit;i++){
          //randomKey gets a random id from the tags
    
          //val is true/false. if val == 1, it's the correct option, if val == 0, it's incorrect.
          if (i===random){
            tag = this.props.tags[id]
            val = 1
            randomKey = id
          }else{
            randomKey = tagKeys.splice(Math.floor((tagKeys.length-1)*Math.random()),1)[0]
            tag = this.props.tags[randomKey]
            val = 0
          }
            options.push(<option value = {val} key = {randomKey} id = {randomKey} onClick = {this.props.checkTagForGame} style= {{fontSize: 14}}>{tag}</option>)
    
    
        }
        return options
      }

      render(){
        let options = this.generateSelectOptionsForGame(this.props.gridnum)
        return(
             <div style={{backgroundColor: "white", opacity: 2, position: "absolute", zLength: 100}}>
                {options}
            </div>
          )
      }

}

export default Options