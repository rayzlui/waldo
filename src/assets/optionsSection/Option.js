import React from 'react';
import PropTypes from 'prop-types';

function checkTagForGame(e, updateScore, resetGrid) {
  //used to check if selected name is correct.
  let val = e.target.value / 1;
  if (val === 1) {
    alert('Correct');
    updateScore(2);
  } else {
    alert('Incorrect');
  }
  resetGrid();
}

export function OptionsView(props) {
  let { gridId, updateScore, currentImage, isTagMode, resetGrid, tags } = props;
  if (isTagMode) return null;
  if (!currentImage || !tags) return null;
  let imageTags = currentImage.tags;
  if (imageTags[gridId] === undefined) return null;
  //just take tags from current image because rest isn't necessary?
  let options = [];
  let tag;
  let val;
  let randomKey;
  let tagKeys = Object.keys(imageTags);
  let limit = tagKeys.length > 4 ? 4 : tagKeys.length; //
  tagKeys.splice(tagKeys.indexOf(gridId.toString()), 1);
  let winningLocation = Math.floor(Math.random() * (limit - 1));
  for (let i = 0; i < limit; i++) {
    if (i === winningLocation) {
      tag = tags[imageTags[gridId]].tag;
      val = 1;
      randomKey = gridId;
    } else {
      //if it's i is not the winningLocation number (which determines where the correct answer should be placed).
      randomKey = tagKeys.splice(
        Math.floor((tagKeys.length - 1) * Math.random()),
        1,
      )[0];
      tag = tags[imageTags[randomKey]].tag;
      val = 0;
    }
    options.push(
      <option
        value={val}
        key={randomKey}
        id={randomKey}
        onClick={event => checkTagForGame(event, updateScore, resetGrid)}
      >
        {tag}
      </option>,
    );
  }
  return options;
}

OptionsView.propTypes = {
  currentImage: PropTypes.object,
  updateScore: PropTypes.func,
  gridId: PropTypes.number,
  isTagMode: PropTypes.bool,
  resetGrid: PropTypes.func,
  tags: PropTypes.object,
};
