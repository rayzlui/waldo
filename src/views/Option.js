import React from 'react';
import PropTypes from 'prop-types';

function checkTagForGame(e, updateScore, resetGrid) {
  //used to check if selected name is correct.
  let val = e.target.value / 1;
  if (val === 1) {
    alert('Correct');
    updateScore(2);
  } else {
    alert('Incorrect');
  }
  resetGrid();
}

export function OptionsView(props) {
  let { gridId, updateScore, currentImage, isTagMode, resetGrid, tags } = props;
  if (isTagMode) return null;
  if (!currentImage || !tags) return null;
  let imageTags = currentImage.tags.reduce((acc, curr) => {
    let key = curr[1];
    let id = curr[0];
    acc[key] = id;
    return acc;
  }, {});
  if (imageTags[gridId] === undefined) return null;
  //just take tags from current image because rest isn't necessary?
  let options = [];
  let tag;
  let val;
  let randomKey;
  //random decides which select option is going to be the correct one.
  //if we're playing game we should ONLY USE KEYS THAT HAVE VALUES.
  let tagKeys = Object.keys(imageTags);
  tagKeys.splice(tagKeys.indexOf(gridId.toString()), 1);
  //remove id from the wrong options to prevent duplicate correct answer.

  let limit = tagKeys.length > 4 ? 4 : tagKeys.length; //

  let winningLocation = Math.floor(Math.random() * limit);
  //random is
  for (let i = 0; i <= limit; i++) {
    //randomKey gets a random id from the tags

    //val is true/false. if val == 1, it's the correct option, if val == 0, it's incorrect.
    if (i === winningLocation) {
      tag = tags[imageTags[gridId]];
      val = 1;
      randomKey = gridId;
    } else {
      //if it's i is not the winningLocation number (which determines where the correct answer should be placed).
      randomKey = tagKeys.splice(
        Math.floor((tagKeys.length - 1) * Math.random()),
        1,
      )[0];
      tag = tags[imageTags[randomKey]];
      val = 0;
    }
    options.push(
      <option
        value={val}
        key={randomKey}
        id={randomKey}
        onClick={event => checkTagForGame(event, updateScore, resetGrid)}
      >
        {tag}
      </option>,
    );
  }
  return options;
}

OptionsView.propTypes = {
  currentImage: PropTypes.object,
  updateScore: PropTypes.func,
  gridId: PropTypes.number,
  isTagMode: PropTypes.bool,
  resetGrid: PropTypes.func,
  tags: PropTypes.array,
};
