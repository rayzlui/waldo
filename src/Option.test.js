import React from 'react'
import Option from './Option.js'
import {shallow} from 'enzyme'

import { isTSAnyKeyword, exportAllDeclaration } from '@babel/types';

/* 
    Since this contains a random generator for correct answer location, 
    we'll use a snapshot to make sure there's a difference between each one.

    Since there is a random generator we'll be testing for the following:
        - There are at most 5 options:
            - We'll test if there are > 5 options and we'll test if there are < 5 options.
        - Test there are no duplicate results aka options.value
        - Test there is only one correct key with (aka one val === 1)
        - Test the correct answer is actually in the options (the value/name of the correct answer.)
        - Make sure click works

*/


describe("Options", ()=>{
    it('should return at most 5 options', ()=>{
        var wrapper1 = shallow(<Option gridnum = {3} tags = {{1:2, 3:4, 5:6, 7:8, 9:10, 11:12}}/>)
        var option1 = wrapper1.find('option')
        expect(option1.length).toEqual(5)
        var wrapper2 = shallow(<Option gridnum = {3} tags = {{1:2, 3:4, 5:6}}/>)
        var option2 = wrapper2.find('option')
        expect(option2.length).toEqual(3)
        var wrapper3 = shallow(<Option gridnum = {3} tags = {{1:2}}/>)
        var option3 = wrapper3.find('option')
        expect(option3.length).toEqual(1)
        var wrapper4 = shallow(<Option gridnum = {3} tags = {{1:2, 3:4, 5:6, 7:8, 9:10, 11:12, 13:14, 15:16, 17:18}}/>)
        var option4 = wrapper4.find('option')
        expect(option4.length).toEqual(5)

    })
    it('should not have duplicate options', ()=>{
        const wrapper = shallow(<Option gridnum = {3} tags = {{1:2, 3:4, 5:6, 7:8, 9:10, 11:12}}/>)
        var options = wrapper.find('option').map(x=>x.debug())
        options.forEach(option=> {
            var removed = options.map(x=>x)
            removed.splice(removed.indexOf(option),1)
            expect(removed).not.toContain(option)
        })
    })

    it('should only have one correct answer and should have a correct answer ', ()=>{
        const wrapper = shallow(<Option gridnum = {3} tags = {{1:2, 3:4, 5:6, 7:8, 9:10, 11:12}}/>)
        var options = wrapper.find('option')
        var values = options.map(x=>x.props().value)
        expect(values.filter(x=>x===1).length).toEqual(1)
    })

    it('should have the right answer with the right value marker (value == 1 and tags[gridnum] should be on the same node)', ()=>{
        var gridnum = 3
        const wrapper = shallow(<Option gridnum = {gridnum} tags = {{1:2, 3:4, 5:6, 7:8, 9:10, 11:12}}/>)
        var options = wrapper.find('option')
        var index = options.map(x=>x.props().value).indexOf(1)
        var randomKey = options.map(x=>x.props().id).indexOf(gridnum)
        expect(index).not.toEqual(-1)
        expect(index).toEqual(randomKey)
    })

    it('should be randomly different each time', ()=>{
        var gridnum = 3
        const wrapper = shallow(<Option gridnum = {gridnum} tags = {{1:2, 3:4, 5:6, 7:8, 9:10, 11:12}}/>)
        expect(wrapper).toMatchSnapshot()
    })

    it('should call function we pass in when clicked', ()=>{
        var gridnum = 3
        var mockFunc = jest.fn()
        const wrapper = shallow(<Option checkTagForGame = {mockFunc}gridnum = {gridnum} tags = {{1:2, 3:4, 5:6, 7:8, 9:10, 11:12}}/>)
        wrapper.find('option').map(x=>x)[0].simulate('click')
        expect(mockFunc).toHaveBeenCalled()
    })



})
