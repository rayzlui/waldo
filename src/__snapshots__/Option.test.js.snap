// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options should be randomly different each time 1`] = `
<div
  style={
    Object {
      "backgroundColor": "white",
      "opacity": 2,
      "position": "absolute",
      "zLength": 100,
    }
  }
>
  <option
    id={1}
    key="1"
    style={
      Object {
        "fontSize": 14,
      }
    }
    value={0}
  >
    2
  </option>
  <option
    id={3}
    key="3"
    style={
      Object {
        "fontSize": 14,
      }
    }
    value={1}
  >
    4
  </option>
  <option
    id={7}
    key="7"
    style={
      Object {
        "fontSize": 14,
      }
    }
    value={0}
  >
    8
  </option>
  <option
    id={9}
    key="9"
    style={
      Object {
        "fontSize": 14,
      }
    }
    value={0}
  >
    10
  </option>
  <option
    id={5}
    key="5"
    style={
      Object {
        "fontSize": 14,
      }
    }
    value={0}
  >
    6
  </option>
</div>
`;
